### PR TITLE
C#: Extract expanded compiler arguments

### PR DESCRIPTION
**Pull request by tamasvajk**

### CODE REVIEW

File: csharp/ql/integration-tests/linux-only/compiler_args/global.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "sdk": {
+        "version": "5.0.408"
```
Comment (tamasvajk) [inline csharp/ql/integration-tests/linux-only/compiler_args/global.json:3]: 🤦

**Comment by tamasvajk** [comment]
`Analysis time, per source` in DCA looks good, at the same time there is some change in `Stage timings`. Are these within the acceptable range?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/Compilation.cs [hunk 2]
```diff
@@ -39,7 +39,35 @@ public override void Populate(TextWriter trapFile)
...
+                if (CommandLineExtensions.IsFileArgument(arg))
+                {
+                    try
+                    {
+                        var rspFileContent = System.IO.File.ReadAllText(arg[1..]);
+                        var rspArgs = CommandLineParser.SplitCommandLineIntoArguments(rspFileContent, removeHashComments: true);
+                        foreach (var rspArg in rspArgs)
+                        {
+                            trapFile.compilation_expanded_args(this, expandedIndex, rspArg);
+                            expandedIndex++;
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/Compilation.cs:57]: I think we tend to write the shorter version elsewhere in the codebase:

```suggestion
                            trapFile.compilation_expanded_args(this, expandedIndex++, rspArg);
```

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/Compilation.cs [hunk 3]
```diff
@@ -39,7 +39,35 @@ public override void Populate(TextWriter trapFile)
...
+                    }
+                    catch (Exception exc)
+                    {
+                        Context.ExtractionError($"Couldn't read compiler argument file: {arg}. {exc.Message}", null, null, exc.StackTrace);
+                    }
+                }
+                else
+                {
+                    trapFile.compilation_expanded_args(this, expandedIndex, arg);
+                    expandedIndex++;
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/Compilation.cs:68]: Same:
```suggestion
                    trapFile.compilation_expanded_args(this, expandedIndex++, arg);
```

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/Compilation.cs [hunk 4]
```diff
@@ -39,7 +39,35 @@ public override void Populate(TextWriter trapFile)
...
+            {
+                var arg = Compilation.Settings.Args[i];
+                trapFile.compilation_args(this, i, arg);
+
+                if (CommandLineExtensions.IsFileArgument(arg))
+                {
+                    try
+                    {
+                        var rspFileContent = System.IO.File.ReadAllText(arg[1..]);
+                        var rspArgs = CommandLineParser.SplitCommandLineIntoArguments(rspFileContent, removeHashComments: true);
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Compilations/Compilation.cs:53]: Nice that we can reuse this method.

**Review by hvitved — APPROVED** [review]
LGTM. A minor suggestion, feel free to apply or not.

File: csharp/ql/integration-tests/linux-only/compiler_args/global.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+    "sdk": {
+        "version": "8.0.100"
```
Comment (michaelnebel) [inline csharp/ql/integration-tests/linux-only/compiler_args/global.json:3]: Depending on the merge order, this PR is relevant: https://github.com/github/codeql/pull/15475

**Comment by michaelnebel** [comment]
> `Analysis time, per source` in DCA looks good, at the same time there is some change in `Stage timings`. Are these within the acceptable range?

Since the overall performance looks good and since we didn't see an overall performance regression, I don't think the change in stage timings matter much, as the measure might be imprecise and since the stats update could have changed the execution order (and maybe stuff that is now faster isn't reported by DCA).

**Review by michaelnebel — APPROVED** [review]
This is excellent!
Thank you for also updating the stats!

**Pull request merged**